src/mpi/pt2pt/recv.c

```diff
@@ -61,6 +61,17 @@ length of the message can be determined with 'MPI_Get_count'.
 @*/
 int MPI_Recv(void *buf, int count, MPI_Datatype datatype, int source, int tag,
              MPI_Comm comm, MPI_Status * status)
+{
+    /* Call the first layer in the QMPI stack that contains the correct function. */
+    for (int i = 0; i <= MPIR_QMPI_num_tools; i++) {
```

Comment:
There's an obvious optimization to be added here when `MPIR_QMPI_num_tools == 0` to avoid the function pointers.